### ISSUE TITLE
Handling large Hypercores persisted to disk (Windows)

### ISSUE THREAD

**Issue by xori** [hypercore-protocol/hypercore#281]
Something I came across while working on the [p2p-indexing-and-search](https://github.com/hypercore-protocol/p2p-indexing-and-search) workshop is that because of how the data is written to disk any hypercore (sparse or otherwise) takes the full space on disk (and then some). Even though solution 14 downloads only __~100kb of data__ from the imdb database it generates __8.92GB of files on disk__.

![image](https://user-images.githubusercontent.com/1091220/100963046-9a93b200-34f3-11eb-85bb-a4b2ab7e9b24.png)
_you can [see the code I use here](https://gist.github.com/xori/3654f3b8cb87744d64524bc62de13059)_

When reading about how hypercore [handles storage](https://datprotocol.github.io/book/ch01-04-storage.html) it seems like it should handle these sparse cases better.

> Because Hypercore and Dat support partially downloading data, a useful feature is to implement sparse persistence. This means that we can write data into memory or to disk with spaces in between, but without paying any cost.

Is this something that needs to be implemented, a regression, or am I just not understanding it properly?

**Comment by xori** [comment]
Upon looking into it more, I presume this is because RAF would need to implement creating these files with the OS's `sparse` flag, which is complicated. I know that in Windows you require Admin rights to create sparse files with `fsutil`, and nodejs's `fs` library doesn't implement the `fallocate` function.

**Comment by xori** [comment]
Surprising no one, actually setting the sparse flag on these files results in huge savings. Still seems huge but I presume that's just Windows trying the best it can.

![image](https://user-images.githubusercontent.com/1091220/100968985-5c50bf80-3500-11eb-8477-92bb22fc42d6.png)

**Comment by xori** [comment]
Set the compress flag, and we hit the size I was expecting.
![image](https://user-images.githubusercontent.com/1091220/100969192-b3569480-3500-11eb-9fc6-6ddc4645826b.png)

**Comment by pfrazee** [comment]
I wrote this a while back so it may be out of date, but here's a random-access storage that's designed for filesystems that can't do sparse indexing: https://github.com/pfrazee/random-access-indexed-file. (This is basically a storage plugin which hypercore can use)

**Comment by mafintosh** [comment]
@xori thanks! Which flags do you set on windows for this? Then I’ll update it. Sorry we aren’t Windows experts

**Comment by xori** [comment]
I only know how to do it via the [`fsutil` command](https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/fsutil-sparse) in Windows.

```powershell
fsutil sparse setflag <file>;  # enables the "<file>" to be marked as "sparse"
fsutil behavior set disablecompression 0 <file>; # by default this is enabled?
```

The downside is that the `fsutil sparse` requires Admin privileges, and I'm unaware of any pure javascript way of marking the file for the OS.

For linux, I only know of `fallocate --dig-holes <file>`, with once again, no pure javascript way of marking this.

It sucks that we have to break out to the OS for this, but thanks @pfrazee for linking me to your RAIF package. I'll monkey patch that into my local hypercore fork for my project for now because I'd rather take the performance hit than hit incompatibility with certain OSs.

**Comment by mafintosh** [comment]
No linux/mac all use sparse files per default

**Comment by xori** [comment]
🤯

**Comment by RangerMauve** [comment]
He he he... feel free to beat me to it with the PR. currently focused on other work and might be a few months off.

I'd be happy to help set it up and integrate with the hyper stack in your apps.

**Comment by mafintosh** [comment]
Fixed in 9.9.0
